### PR TITLE
Split registergui into Widget and Dialog

### PR DESCRIPTION
**Pull request by alikins**
Implement RegisterWidget flow with proceed and finished signals

Make SubManBaseWidget also a GObject.GObject. So that
sub classes of it can use gobject signals and properties.

Add 'proceed' signal to RegisterWidget, and a handler that
will move it to the next screen. To drive the widget, another
widget that provides cancel/register buttons makes them
emit the 'proceed' signal on registerWidget.

RegisterWidget also emits a 'finished' signal when registration
is done. The parent widget needs to deal with that. The default
RegisterDialog just hides the register window.

Try making RegisterInfo a GObject with properties
Connect RegisterInfo's 'notify::username' and password
signals to callback that updates the cp_provider

Start splitting error handling up and using signals

ie, register error, register failure, attach error, attach
failure raise signals so they can get different handlers, depending
on the dialog that holds the register widget.

add RegisterDialog handlers for attach/register error and failure

Registration state (as used in global STATE) is now
a property on the registration widget, and the labels/titles
that need to be updated when it changes are hooked to the
notify signal on RegisterWidget's 'register_state' property.

Ditto for RegisterWidget's 'details-label-txt', ie, the
txt for the label under the progress bar ('Finding suitable service
levels', etc).

There is now a 'stay-on-screen' signal raise from applies or
async callbacks to indicate to stay on the current screen, replacing
the previous use of sending the widget to the 'DONT_CHANGE' screen.

Stop using handle_gui_exception for non-exceptions

Passing None as the 'exception' eventually causes
the last exception handled in the main context to
get logged as an exception, and also prevents the
dialog from displaying the message, and generally
confusing the event loop.

Also update SelectSLAScreen to use a tuple of
(sla_name, sla_map) in its model for the set of
radio buttons, to avoid referncing the
now removed self._sla_data_map

Prevent gui from disabling act keys to prod.

Split handle_gui_exception msg formatting into methods

Start emitting error signals from RegisterWidget with msgs

rhsm_gui can use initial-setup spoke's set_error(msg) to show
errors and failures.

Screen and NoGuiScreen, RegisterWidget, and RegisterDialog
all know of 'register-error' signal.

Screens can emit it, where RegisterWidget catches it, logs
it, stops emission of that signal. (Either the screens
or RegisterWidget could provide a class closure do_register_error
that that would see the signal first, and that can be used
to move to approriate screen to fix the error).
RegisterWidget then emits RegisterWidget's
'register-error' signal, that RegisterDialog is connected to.

RegisterDialog._on_register_error will call
RegisterDialogs handle_register_error or
handle_register_exception

Any errors that need to present a dialog or other info
now need to emit a 'register-error' with the args
of (msg, exc_info). exc_info can be None.

### CODE REVIEW

**Comment by alikins** [comment]
Build 'Ro' Bot: retest this please

**Comment by cnsnyder** [comment]
@alikins 
When I click on the "Configure Proxy" button during initial-setup I receive this traceback

Traceback (most recent call last):
  File "/usr/share/rhsm/subscription_manager/gui/registergui.py", line 1324, in _on_proxy_button_clicked
    self.network_config_dialog.set_parent_window(self.parent_window)
  File "/usr/share/rhsm/subscription_manager/gui/networkConfig.py", line 310, in set_parent_window
    self.networkConfigDialog.set_transient_for(window)
TypeError: argument parent: Expected Gtk.Window, but got gi.repository.AnacondaWidgets.SpokeWindow

**Comment by cnsnyder** [comment]
Clicking the "FINISH CONFIGURATION" button, after completing registration and accepting the license agreement caused the following:

Exception AttributeError: "'NoneType' object has no attribute 'udev_unref'" in <bound method Context.__del__ of <pyudev.core.Context object at 0x2cac810>> ignored
[root@localhost subscription-manager]#

**Comment by cnsnyder** [comment]
This is minor as it seems to be purely cosmetic, but the bouncing loading bar during attach seems to bounce faster in initial-setup than in the regular gui.

**Comment by alikins** [comment]
Exception AttributeError: "'NoneType' object has no attribute 'udev_unref'" in > ignored
[root@localhost subscription-manager]# 

As far as I can tell, that's unrelated to our code. It happens without our modules.

**Comment by alikins** [comment]
The progress bar thing perplexes me. It is definitely very slow in rhel7 subscription-manager-gui, especially against stage. Oddly, the same code works as expected on RHEL6, and f22. It's odd
because it is driven by a timeout so the env should not affect it at all.

I think it may be related to the way the gui/registration/AsyncBackend._watch_thread adds idle handlers, but can't put a finger on why (especially when it works newer and older)

**Comment by cnsnyder** [comment]
I get different results for auto-attach in initial-setup and the regular gui.
In the regular gui installed products lists:
"Red Hat Enterprise Linux Server"

with service levels:
"Premium
Self-Support"

On initial-setup the installed products list appears empty.
The service levels listed for initial-setup:
"Standard
Premium
None
Self-Support"

File: src/subscription_manager/gui/registergui.py [hunk 1]
```diff
@@ -122,388 +117,528 @@ def reset_resolver():
...
-    def _get_initial_screen(self):
-        return SELECT_SLA_PAGE
+        super(AutobindWizardDialog, self).show()
+        self.register_widget.change_screen(SELECT_SLA_PAGE)
 
 
+# TODO: Screen could be a container widget, that has the rest of the gui as
+#       a child. That way, we could add the Screen class to the
+#       register_notebook directly, and follow up to the parent the normal
+#       way. Then we could stop passing 'parent' around. And RegsiterInfo
```
Comment (cnsnyder) [inline src/subscription_manager/gui/registergui.py]: Typo: 'RegsiterInfo' -> 'RegisterInfo'

**Comment by cnsnyder** [comment]
Looks good, ACK!

**Pull request merged**